src/Microsoft.AspNet.PlatformHandler/HttpPlatformHandlerMiddleware.cs

```diff
@@ -0,0 +1,85 @@
+using Microsoft.Framework.Primitives;
+
+namespace Microsoft.AspNet.PlatformHandler
+{
+    public class HttpPlatformHandlerMiddleware
+    {
+        private const string XForwardedForHeaderName = "X-Forwarded-For";
+        private const string XForwardedProtoHeaderName = "X-Forwarded-Proto";
+        private const string XIISWindowsAuthToken = "X-IIS-WindowsAuthToken";
+        private const string XOriginalProtoName = "X-Original-Proto";
+        private const string XOriginalIPName = "X-Original-IP";
+
+        private readonly RequestDelegate _next;
+
+        public HttpPlatformHandlerMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            var xForwardProtoHeaderValue = httpContext.Request.Headers[XForwardedProtoHeaderName];
+            if (!string.IsNullOrEmpty(xForwardProtoHeaderValue))
+            {
+                if (!string.IsNullOrEmpty(httpContext.Request.Scheme))
+                {
+                    httpContext.Request.Headers[XOriginalProtoName] = httpContext.Request.Scheme;
+                }
+                httpContext.Request.Scheme = xForwardProtoHeaderValue;
+            }
+            
+            var xForwardedForHeaderValue = httpContext.Request.Headers.GetCommaSeparatedValues(XForwardedForHeaderName);
+            if (xForwardedForHeaderValue != null && xForwardedForHeaderValue.Length > 0)
+            {
+                IPAddress ipFromHeader;
+                if (IPAddress.TryParse(xForwardedForHeaderValue[0], out ipFromHeader))
+                {
+                    var remoteIPString = httpContext.Connection.RemoteIpAddress?.ToString();
+                    if (!string.IsNullOrEmpty(remoteIPString))
+                    {
+                        httpContext.Request.Headers[XOriginalIPName] = remoteIPString;
+                    }
+                    httpContext.Connection.RemoteIpAddress = ipFromHeader;
+                }
+            }
+
+            var xIISWindowsAuthToken = httpContext.Request.Headers[XIISWindowsAuthToken];
+            int hexHandle;
+            if (!StringValues.IsNullOrEmpty(xIISWindowsAuthToken)
+                && int.TryParse(xIISWindowsAuthToken, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out hexHandle))
+            {
+                var handle = new IntPtr(hexHandle);
+                var winIdentity = new WindowsIdentity(handle);
+                // WindowsIdentity just duplicated the handle so we need to close the original.
+                NativeMethods.CloseHandle(handle);
+
+                httpContext.Response.RegisterForDispose(winIdentity);
+                var winPrincipal = new WindowsPrincipal(winIdentity);
+
+                var existingPrincipal = httpContext.User;
+                if (existingPrincipal != null)
+                {
+                    httpContext.User = SecurityHelper.MergeUserPrincipal(existingPrincipal, winPrincipal);
```

Comment:
I believe this thing handles nulls? https://github.com/aspnet/Common/blob/dev/src/Microsoft.Framework.SecurityHelper.Sources/SecurityHelper.cs#L21

So no need to check for null. Unless you're trying to avoid the ClaimsPrincipal wrapping the WindowsIdentity?